src/libtiled/mapwriter.cpp

```diff
@@ -618,74 +652,100 @@ void MapWriterPrivate::writeObjectGroup(QXmlStreamWriter &w,
 {
     w.writeStartElement(QLatin1String("object"));
     const int id = mapObject.id();
     const QString &name = mapObject.name();
     const QString &type = mapObject.type();
-    if (id != 0)
+
+    const QPointF pos = mapObject.position();
+
+    TemplateRef templateRef = mapObject.templateRef();
+    const int templateId = templateRef.templateId;
+    const auto group = templateRef.templateGroup;
+    bool isTemplateInstance = group;
+
+    if (isTemplateInstance) {
+        int groupFirstTid = mTidMapper.templateGroupToFirstTid(group);
+        int tid = templateId + groupFirstTid;
+        w.writeAttribute(QLatin1String("tid"), QString::number(tid));
+    }
+
+    if (id != 0 && !isTemplateInstance)
         w.writeAttribute(QLatin1String("id"), QString::number(id));
-    if (!name.isEmpty())
+
+    if (shouldWrite(!name.isEmpty(), isTemplateInstance, mapObject.propertyChanged(MapObject::NameProperty)))
         w.writeAttribute(QLatin1String("name"), name);
-    if (!type.isEmpty())
+
+    if (shouldWrite(!type.isEmpty(), isTemplateInstance, mapObject.propertyChanged(MapObject::TypeProperty)))
         w.writeAttribute(QLatin1String("type"), type);
 
-    if (!mapObject.cell().isEmpty()) {
+    if (shouldWrite(!mapObject.cell().isEmpty(), isTemplateInstance, mapObject.propertyChanged(MapObject::CellProperty))) {
         const unsigned gid = mGidMapper.cellToGid(mapObject.cell());
         w.writeAttribute(QLatin1String("gid"), QString::number(gid));
     }
 
-    const QPointF pos = mapObject.position();
-    const QSizeF size = mapObject.size();
-
     if (id != 0) {
         w.writeAttribute(QLatin1String("x"), QString::number(pos.x()));
         w.writeAttribute(QLatin1String("y"), QString::number(pos.y()));
     }
 
-    if (size.width() != 0)
-        w.writeAttribute(QLatin1String("width"), QString::number(size.width()));
-    if (size.height() != 0)
-        w.writeAttribute(QLatin1String("height"), QString::number(size.height()));
+    if (shouldWrite(true, isTemplateInstance, mapObject.propertyChanged(MapObject::SizeProperty))) {
+        const QSizeF size = mapObject.size();
+        if (size.width() != 0)
+            w.writeAttribute(QLatin1String("width"), QString::number(size.width()));
+        if (size.height() != 0)
+            w.writeAttribute(QLatin1String("height"), QString::number(size.height()));
+    }
 
     const qreal rotation = mapObject.rotation();
-    if (rotation != 0.0)
+    if (shouldWrite(rotation != 0.0, isTemplateInstance, mapObject.propertyChanged(MapObject::RotationProperty)))
         w.writeAttribute(QLatin1String("rotation"), QString::number(rotation));
 
-    if (!mapObject.isVisible())
+    if (shouldWrite(!mapObject.isVisible(), isTemplateInstance, mapObject.propertyChanged(MapObject::VisibleProperty)))
         w.writeAttribute(QLatin1String("visible"), QLatin1String("0"));
 
-    writeProperties(w, mapObject.properties());
+    if (!isTemplateInstance)
```

Comment:
Properties should always be written, since this will only include overridden values and any properties only added to the instance.